data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,689 @@
+```ocaml
+# raise;;
+- : exn -> 'a' = <fun>
+```
+
+Functions may have several parameters.
+```ocaml
+# fun a b -> a ^ " " ^ b;;
+- : string -> string -> string = <fun>
+
+# let mean a b = (a + b) / 2;;
+val mean : int -> int -> int = <fun>
+```
+
+As for the product type symbol `*`, the function type symbol `->` is not associative. These two types are not the same.
+- `(int -> int) -> int` : This is a function taking a function of type `int -> int` as parameter and returning an `int` as result.
+- `int -> (int -> int)` : This is a function taking an `int` as parameter and returning a function of type `int -> int` as result.
+
+### Unit
+
+A unique value has type `unit`. It is written `()` and pronounced “unit”.
+
+The `unit` type has several usages. One of its main roles is to serve as a token when a function does not need to be passed data or doesn't have any data to return once it has completed its computation. This happens when functions have side effects such as OS-level I/O. Functions need to be applied to something for their computation to be triggered, and they also must return something. When nothing making sense can be passed or returned, `()` should be used.
+```ocaml
+# read_line;;
+- : unit -> string = <fun>
+
+# print_endline;;
+- : string -> unit = <fun>
+```
+
+Function `read_line` reads an end-of-line terminated sequence of characters from standard input and returns it as a string. Reading input begins when `()` is passed.
+
+Function `print_endline` prints the string followed by a line ending on standard output. Return of the unit value means the output request has been queued by the operating system.
+
+## User-Defined Types
+
+### Variants
+
+#### Enumerated Data Types
+
+The simplest form of a variant type corresponds to an [enumerated type](https://en.wikipedia.org/wiki/Enumerated_type). It is defined by an explicit list of named values. Defined values are called constructors and must be capitalised.
+
+For example, here is how variant data types could be defined to represent Dungeons & Dragons character classes and alignment.
+```ocaml
+# type character_class =
+    | Barbarian
+    | Bard
+    | Cleric
+    | Druid
+    | Fighter
+    | Monk
+    | Paladin
+    | Ranger
+    | Rogue
+    | Sorcerer
+    | Wizard;;
+type character_class =
+    Barbarian
+  | Bard
+  | Cleric
+  | Druid
+  | Fighter
+  | Monk
+  | Paladin
+  | Ranger
+  | Rogue
+  | Sorcerer
+  | Wizard
+
+# type rectitude = Evil | R_Neutral | Good;;
+type rectitude = Evil | R_Neutral | Good
+
+# type firmness = Chaotic | F_Neutral | Lawful;;
+type firmness = Chaotic | F_Neutral | Lawful
+```
+
+These kinds of variant types can also be used to represent weekdays, cardinal
+directions, or any other fixed-sized set of values that can be given names. A
+total ordering is defined on values following the definition order (e.g., `Druid
```

Comment:
Should "total order" be explained? I don't know if it's reasonable to assume readers will know what it means otherwise.